gobblin-metastore/src/main/java/gobblin/metastore/DatabaseJobHistoryStore.java

```diff
@@ -318,296 +271,440 @@ public void close()
+    upsertStatement.setString(++index,
+        info.hasTable() && info.getTable().hasNamespace() ? info.getTable().getNamespace() : null);
+    upsertStatement.setString(++index, info.hasTable() && info.getTable().hasName() ? info.getTable().getName() : null);
+    upsertStatement.setString(++index,
+        info.hasTable() && info.getTable().hasType() ? info.getTable().getType().name() : null);
+    upsertStatement.addBatch();
   }
 
-  private void updateProperty(Connection connection, String template, String id, String key, String value,
-      boolean insert)
-      throws SQLException {
-    Preconditions.checkArgument(!Strings.isNullOrEmpty(id));
-    Preconditions.checkArgument(!Strings.isNullOrEmpty(key));
-    Preconditions.checkArgument(!Strings.isNullOrEmpty(value));
+  private List<JobExecutionInfo> processQueryByIds(Connection connection, JobExecutionQuery query,
+                                                         Optional<String> tableFilter, List<String> jobIds)
+          throws SQLException {
+    Map<String, JobExecutionInfo> jobExecutionInfos = getJobExecutionInfos(connection, jobIds);
+    addMetricsToJobExecutions(connection, query, jobExecutionInfos);
+    addPropertiesToJobExecutions(connection, query, jobExecutionInfos);
+    addTasksToJobExecutions(connection, query, tableFilter, jobExecutionInfos);
 
-    PreparedStatement updateStatement = connection.prepareStatement(template);
-    int index = 0;
-    if (insert) {
-      updateStatement.setString(++index, id);
-      updateStatement.setString(++index, key);
-      updateStatement.setString(++index, value);
-    } else {
-      updateStatement.setString(++index, value);
-      updateStatement.setString(++index, id);
-      updateStatement.setString(++index, key);
-    }
-    updateStatement.executeUpdate();
+    return ImmutableList.copyOf(jobExecutionInfos.values());
   }
 
-  private JobExecutionInfo processQueryById(Connection connection, String jobId, JobExecutionQuery query,
-      Optional<String> tableFilter) throws SQLException {
-    Preconditions.checkArgument(!Strings.isNullOrEmpty(jobId));
-
-    // Query job execution information
-    PreparedStatement jobExecutionQueryStatement =
-        connection.prepareStatement(JOB_EXECUTION_QUERY_BY_JOB_ID_STATEMENT_TEMPLATE);
-    jobExecutionQueryStatement.setString(1, jobId);
-    ResultSet jobRs = jobExecutionQueryStatement.executeQuery();
-    if (!jobRs.next()) {
-      return null;
+  private Map<String, JobExecutionInfo> getJobExecutionInfos(Connection connection, List<String> jobIds) throws SQLException {
+    Map<String, JobExecutionInfo> jobExecutionInfos = Maps.newLinkedHashMap();
+    if (jobIds != null && jobIds.size() > 0) {
+      String template = String.format(JOB_EXECUTION_QUERY_BY_JOB_ID_STATEMENT_TEMPLATE,
+              StringUtils.join(Iterables.limit(Iterables.cycle("?"), jobIds.size()).iterator(), ","));
```

Comment:
This pattern seems to be duplicate in a few places, can it be moved to its own method.